### DISCUSSION TITLE
Read multi part CRYPTO-PSBT data

### DISCUSSION THREAD

**Discussion by vemla-solutions** [BlockchainCommons/Gordian-Developer-Community#68] [Q&A]
Could you please advise or reference to a description of how to read the multipart crypto-psbt data?

When I get an animated QR code displayed, I get a sequence of parts like:

UR:CRYPTO-PSBT/195-7/LPCSSRATCSSGCYWZJOUORDHDCAGASWLEKKFYYLWNLGLNMDWYFGBYLELFSRMKDPDLWYUYOSEOJTIOADAEAEAEKKOEVYPD

UR:CRYPTO-PSBT/196-7/LPCSSSATCSSGCYWZJOUORDHDCALUMSDPOXIECYLAENGSRKKOVWMNRLHFHENYJSHDPDBNHGBYTSNYZSTKZCTDHEYKVYKK

UR:CRYPTO-PSBT/197-7/LPCSSKATCSSGCYWZJOUORDHDCAWZJLHLBZFXETOEMUDYFSYNRKEEWSFNHDSWYTFHNLTALTHYBSAHURBYVWKGRDNEJNKI

etc.

But how to join them into 1 transaction? I had an impression that there are 7 parts for this one, and then the same 7 parts should be rotated in sequential order, e.g. 196-7 would be the first part, because mod(196,7) = 0 and then 7 parts are displayed in sequential order, but it looks like it is not the case. I see there are some parts repeating, but they are not on 7 cycle and also are displayed quite randomly in my opinion.

I checked documentation and for PSBT there is no clear description present, e.g.
https://github.com/BlockchainCommons/crypto-commons/blob/master/Docs/ur-1-overview.md
says "[A Guide to Using URs for PSBTs] (pending)" and no link.

What I reached so far is bytewords decoding each part, and then CBOR decoding each part and then I get the following for the first part, for example:

LPCSSRATCSSGCYWZJOUORDHDCAGASWLEKKFYYLWNLGLNMDWYFGBYLELFSRMKDPDLWYUYOSEOJTIOADAEAEAEKKOEVYPD
bytewords decoded:
8518C30718CA1AF270DCBA581D49C68A7944F7F18D8695EE46118A82C3982D2FEEDBA7336E670100000079A2E1A8
CBOR decoded:
[195, 7, 202, 4067482810, h'49C68A7944F7F18D8695EE46118A82C3982D2FEEDBA7336E6701000000']

195, 7 and the partial message at the end are easy to understand, but what are 202 and 4067482810? They seem to be repeating in each part as well. Should they be the hints of how to glue the parts and the order to select?

I will appreciate any guidance here.

Also if you can share any guidance of how to get a transaction that can be broadcasted once all parts are glued, will be also helpful. Or will it be just glued hex-coded transaction that can be immediately used for the network broadcast?

Thank you!

**Comment by wolfmcnally** [comment]
Please watch the [Blockchain Commons Tech Overview](https://www.youtube.com/watch?v=RYgOFSdUqWY) video. It will answer many of the questions you just asked.

**Reply by vemla-solutions** [reply]
Hi @wolfmcnally 
Thank you for this video link, it is a good overview, however it misses exactly the details that I look for.
It has some piece of info at 23 minute and later, that when UR code is split into parts, the first 7 are all the needed parts, however, if the animated QR code continues the rest are pseudo-randomly displayed parts and the receiver of the QR code need to apply the same pseudo random logic to pull the data. But is this logic described anywhere?
I want to create a mobile application that will be able to scan psbt and broadcast to the network, so I'm missing the description of how to read the data from the animated QR code for this UR type (crypto-psbt), is it specified anywhere? I can't find myself, unfortunately.
Thank you in advance!

**Comment by ChristopherA** [comment]
Have you looked over our [UR Introduction for Developers](https://github.com/BlockchainCommons/crypto-commons/blob/master/Docs/ur-1-overview.md)?

There is also a mentions a list of third-party libraries that support UR in a variety of languages (right now Java, Python & Rust in addition to the reference C++ implementation) at https://github.com/BlockchainCommons/crypto-commons#bc-ur — I also think there is a Javascript implementation out there, but it doesn't seem to be in that list (any lurkers know where it is?)

I hope that one of these helps solves your problem.

-- Christopher Allen

**Reply by Overtorment** [reply]
https://github.com/ngraveio/bc-ur
https://github.com/KeystoneHQ/ur-registry

**Reply by ChristopherA** [reply]
Thanks! @shannona we should add their implementation to the list.

**Reply by vemla-solutions** [reply]
Thank you @ChristopherA and @Overtorment 
I checked some other languages implementations before and looked into JS now.
I use flutter / Dart, unfortunately, I could not find any UR implementation there, so probably will need to write from scratch, at least the basic mechanisms.

I found useful the following link useful: https://github.com/BlockchainCommons/Research/blob/master/papers/bcr-2020-005-ur.md#ur-encoding

As now I have an understanding about the logic, however, still struggle to go through all the steps with a real example (I scanned animated QR codes from both Sparrow wallet and Seedsigner).

I understand that this is too much to ask, but it would be much appreciated if you could add a step by step example in the description in the future (encode and decode). I think it would help many other developers with implementation.

Thanks a lot for your help, for now I will try to proceed with own research and attempts to solve it.

**Reply by wolfmcnally** [reply]
> I understand that this is too much to ask, but it would be much appreciated if you could add a step by step example in the description in the future (encode and decode). I think it would help many other developers with implementation.

If you're going to do a new implementation, I suggest you follow the same path I have recommended to other developers: focus on the UR unit tests and write the code that produces the same result as those tests. In my own code I have arranged the tests so that later tests rely on pieces tested by earlier tests. So if you translate the code in the order that makes the tests run, you should be on a path to success.

**Comment by shannona** [comment]
@vemla-solutions Please also see:
https://github.com/BlockchainCommons/crypto-commons/blob/master/Docs/ur-1-overview.md

(and the linked docs)

If there's something more you think is necessary in those, let me know.
